editor/src/core/tailwind/tailwind-compilation.ts

```diff
@@ -0,0 +1,110 @@
+import React from 'react'
+import type { TailwindConfig, Tailwindcss } from '@mhsdesign/jit-browser-tailwindcss'
+import { createTailwindcss } from '@mhsdesign/jit-browser-tailwindcss'
+import type { ProjectContentTreeRoot } from 'utopia-shared/src/types'
+import { getProjectFileByFilePath, walkContentsTree } from '../../components/assets'
+import { interactionSessionIsActive } from '../../components/canvas/canvas-strategies/interaction-state'
+import { CanvasContainerID } from '../../components/canvas/canvas-types'
+import {
+  Substores,
+  useEditorState,
+  useRefEditorState,
+} from '../../components/editor/store/store-hook'
+import { importDefault } from '../es-modules/commonjs-interop'
+import { rescopeCSSToTargetCanvasOnly } from '../shared/css-utils'
+import type { RequireFn } from '../shared/npm-dependency-types'
+import { TailwindConfigPath } from './tailwind-config'
+import { memoize } from '../shared/memoize'
+import { equalityFnForProjectContents } from '../shared/project-contents-utils'
+
+const TAILWIND_INSTANCE: { current: Tailwindcss | null } = { current: null }
+
+export function isTailwindEnabled(): boolean {
+  return TAILWIND_INSTANCE.current != null
+}
+
+function ensureElementExists({ type, id }: { type: string; id: string }) {
+  let tag = document.getElementById(id)
+  if (tag == null) {
+    tag = document.createElement(type)
+    tag.id = id
+    document.head.appendChild(tag)
+  }
+  return tag
+}
+
+async function generateTailwindStyles(tailwindCss: Tailwindcss, allCSSFiles: string) {
+  const contentElement = document.getElementById(CanvasContainerID)
+  if (contentElement == null) {
+    return
+  }
+  const content = contentElement.outerHTML
+  const styleString = await tailwindCss.generateStylesFromContent(allCSSFiles, [content])
```

Comment:
what does allCSSFiles do here? is generateStylesFromContent using it to avoid duplicates, or is it appending them all to the styleString? if the latter, does it mean if I have any css anywhere in the project, it will be attached to the DOM? that's certainly not something desired

Reply:
Context on what `tailwindCss.generateStylesFromContent` does:

Tailwind styles are generated based on the tailwind config (passed to the constructor of the `tailwindCss` instance), the css files (that contain the `@tailwind` directives, custom font faces, [layers](https://tailwindcss.com/docs/adding-custom-styles#using-css-and-layer)), and the actual tailwind classes in the elements we're interested in (the second argument of `generateStylesFromContent`). Based on all this, `generateStylesFromContent` returns the CSS generated by the Tailwind JIT compiler, which has been optimized for the given content.

Specific answers:
- `generateStylesFromContent` is not using `allCSSFiles` to avoid duplicates
- I haven't specifically debugged the output of `generateStylesFromContent`, since it's been treated as a black box so far

Reply:
thanks! how does "real" tailwind ingest these @tailwind directives? if I write a directive in a css file that I never reference anywhere, will it be picked up by the tailwind/scss precompile step in a real application? As far as I understand, in this implementation here, it _will_ be picked up automatically. I just want to make sure if that matches the real tailwind implementation

Reply:
Tailwind is a used as a PostCSS plugin, and the files that PostCSS processes are determined by the build system (see examples here: https://github.com/postcss/postcss?tab=readme-ov-file#usage). Replicating that functionality is out of scope for this PR, we'll polish this on a follow-up PR